File: LeetCode/135Candy/Candy.py
class Solution:
    def candy(self, ratings: List[int]) -> int:
        # Time:O(n), Space:O(n)
        n = len(ratings)
        l2r = [1 for i in range(n)]
        r2l = [1 for i in range(n)]
        candies = 0
        for i in range(n):
            if i > 0 and ratings[i] > ratings[i-1]:
                l2r[i] = l2r[i-1] + 1
        for i in range(n-2, -1, -1):
            if i > -1 and ratings[i] > ratings[i+1]:
                r2l[i] = r2l[i+1] + 1
        for i in range(n):
            candies += max(l2r[i], r2l[i])
        return candies

File: LeetCode/135Candy/Candy.cpp
class Solution {
public:
    int candy(vector<int>& ratings) {
        // Time:O(n), Space:O(1)
        int n = ratings.size(), candies = 0;
        std::vector<int> l2r(n, 1);
        std::vector<int> r2l(n, 1);
        for (int i = 0; i < n; i++) {
            if (i > 0 && ratings[i] > ratings[i-1]) {
                l2r[i] = l2r[i-1] + 1;
            }
        }
        for (int i = n-2; i > -1; i--) {
            if (i > -1 && ratings[i] > ratings[i+1]) {
                r2l[i] = r2l[i+1] + 1;
            }
        }
        for (int i = 0; i < n; i++) {
            if (l2r[i] > r2l[i]) {
                candies += l2r[i];
            }
            else {
                candies += r2l[i];
            }
        }
        return candies;
    }
};
